panther_hardware_interfaces/src/panther_system.cpp

```diff
@@ -0,0 +1,519 @@
+    }
+
+  } catch (const std::runtime_error & e) {
+    roboteq_error_filter_->UpdateError(static_cast<std::size_t>(ErrorsFilterIds::READ_PDO), true);
+    RCLCPP_ERROR_STREAM(logger_, "Error when trying to read feedback: " << e.what());
+  }
+}
+
+void PantherSystem::UpdateMsgErrors()
+{
+  CanErrors can_errors;
+  can_errors.error = roboteq_error_filter_->IsError();
+  can_errors.write_sdo_error =
+    roboteq_error_filter_->IsError(static_cast<std::size_t>(ErrorsFilterIds::WRITE_SDO));
+  can_errors.read_sdo_error =
+    roboteq_error_filter_->IsError(static_cast<std::size_t>(ErrorsFilterIds::READ_SDO));
+  can_errors.read_pdo_error =
+    roboteq_error_filter_->IsError(static_cast<std::size_t>(ErrorsFilterIds::READ_PDO));
+  can_errors.front_data_timed_out = motors_controller_->GetFrontData().IsDataTimedOut();
+  can_errors.rear_data_timed_out = motors_controller_->GetRearData().IsDataTimedOut();
+  can_errors.front_can_net_err = motors_controller_->GetFrontData().IsCanNetErr();
+  can_errors.rear_can_net_err = motors_controller_->GetRearData().IsCanNetErr();
+
+  panther_system_ros_interface_.UpdateMsgErrors(can_errors);
+}
+
+return_type PantherSystem::read(
+  const rclcpp::Time & /* time */, const rclcpp::Duration & /* period */)
+{
+  UpdateDriverState();
+  UpdateSystemFeedback();
+  UpdateMsgErrors();
+
+  panther_system_ros_interface_.PublishDriverState();
+
+  return return_type::OK;
+}
+
+return_type PantherSystem::write(
+  const rclcpp::Time & /* time */, const rclcpp::Duration & /* period */)
+{
+  if (!roboteq_error_filter_->IsError()) {
+    try {
+      motors_controller_->WriteSpeed(
+        hw_commands_velocities_[0], hw_commands_velocities_[1], hw_commands_velocities_[2],
+        hw_commands_velocities_[3]);
+      roboteq_error_filter_->UpdateError(
+        static_cast<std::size_t>(ErrorsFilterIds::WRITE_SDO), false);
+    } catch (const std::runtime_error & e) {
+      RCLCPP_ERROR_STREAM(logger_, "Error when trying to write commands: " << e.what());
+      roboteq_error_filter_->UpdateError(
+        static_cast<std::size_t>(ErrorsFilterIds::WRITE_SDO), true);
+    }
+  }
+
+  // "soft" error - still there is communication over CAN with drivers, so publishing feedback is
+  // continued - hardware interface's onError isn't triggered estop is handled similarly - at the
+  // time of writing there wasn't a better approach to handling estop. Moved after WriteSpeed, so
+  // that safety stop can be set just after write error happens.
+  if (roboteq_error_filter_->IsError()) {
+    if (
+      (motors_controller_->GetFrontData().GetLeftRuntimeError().GetMessage().safety_stop_active &&
+       motors_controller_->GetFrontData().GetRightRuntimeError().GetMessage().safety_stop_active &&
+       motors_controller_->GetRearData().GetLeftRuntimeError().GetMessage().safety_stop_active &&
+       motors_controller_->GetRearData().GetRightRuntimeError().GetMessage().safety_stop_active) ==
+      false) {
```

Comment:
```suggestion
    if (
      !motors_controller_->GetFrontData().GetLeftRuntimeError().GetMessage().safety_stop_active ||
      !motors_controller_->GetFrontData().GetRightRuntimeError().GetMessage().safety_stop_active ||
      !motors_controller_->GetRearData().GetLeftRuntimeError().GetMessage().safety_stop_active ||
      !motors_controller_->GetRearData().GetRightRuntimeError().GetMessage().safety_stop_active) {
```
Why not like this? 
Or even:
```cpp
    if (!SafetyStopActive()) {
    .
    .
    .
bool PantherSystem::SafewtyStopActive()
{
  return motors_controller_->GetFrontData().GetLeftRuntimeError().GetMessage().safety_stop_active &&
         motors_controller_->GetFrontData()
           .GetRightRuntimeError()
           .GetMessage()
           .safety_stop_active &&
         motors_controller_->GetRearData().GetLeftRuntimeError().GetMessage().safety_stop_active &&
         motors_controller_->GetRearData().GetRightRuntimeError().GetMessage().safety_stop_active;
}
```